rfcs/records.md

```diff
@@ -0,0 +1,450 @@
+    return Point(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum()
+    return self.x + self.y
+end
+
+function Point.newDiagonal(v)
+    return Point(v, v)
+end
+```
+
+Note that `Point` is simply a table and as such it can be used to store static methods as well; as it also serves as a metatable, metafields defined on this table
+will change the behavior of the record values. The shape isn't frozen automatically but can be frozen manually if desired via `table.freeze`.
+
+To create a record, you need to use a record constructor. This is done using call-like syntax:
+
+Syntax A:
+
+```lua
+local person = Person { name = "Bob", age = 42 }
+```
+
+Syntax B:
+
+```lua
+local person = Person("Bob", 42)
+```
+
+The big difference between variants A and B is whether you need to spell out field names at construction time. There's precedents for going either way; some
+languages like F#/C#/Kotlin implement record construction as a function call and when you define a record, you essentially define the record constructor. This
+is beautifully concise, but is a bit more difficult to migrate away from tables, and it's easy to mix up the names. Variant A is more verbose and requires a bit
+more magic at compile time to keep construction efficient.
+
+In variant A, it would make sense to allow omission of any field, in which case it gets replaced with the default of `nil`. A future extension (not part of this RFC) could be made to
+allow specification of default values at record definition time. Type checker would fail to type check record construction if fields that have non-optional types
+have the values omitted.
+
+In variant B, it would probably make sense to require exact number of values to be specified, or follow the usual function call syntax rules.
+
+Note that since records are first class objects, you can export or import a record through a module boundary in the usual way:
+
+```lua
+local HR = require(path)
+local r = HR.Person { name = "Bob", age = 42 } -- or HR.Person(1, 2) in variant B
+```
+
+### Generic records
+
+At definition point, records can have generic arguments that can be used in the field type specification:
+
+```lua
+record Point<V> = { x: V, y: V }
+```
+
+When record names are used in type context, they use the standard generic instantiation syntax to specify the generic parameters:
+
+```lua
+local p: Point<number>
+```
+
+When record names are used in record literals, they don't specify the generic parameters. This is to avoid complexity with parsing `<` in expression context:
+
+```lua
+local p: Point<number> = Point { x = 1, y = 2 }
+```
+
+The generic type parameters are erased at runtime.
+
+### Type checking records
+
+Records defined via a `record` statement can be used in type annotations.
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+> Alternatively, is `export type Record = Record` too awkward?
```

Comment:
Too awkward, painful, and error prone. If your records are generic, you will have to explicitly spell out the type parameters and then correctly pass in the type arguments of those parameters. Also, the type arguments could be out of order by accident.

While plans around bounded quantification is not being formed at this time, it's possible that you need to duplicate your generic bounds twice, once in your record definition and once again for the same corresponding type parameter in the exported type alias. That's if we decide due to some cases that Luau shouldn't infer generic bounds for explicit generics.

Then there's the matter of scoping, `type Foo = Foo` will circularly reference itself, not the pre-existing type definition.

There's two ways I can see out of this:
1. Go with `export record` syntax. It does leave a question of what to do about the record _value_. We don't have precedent yet about exporting values. Perhaps we should define that first in a separate RFC and have that RFC be a blocker for this?
2. Come up with a re-export syntax dedicated for type definitions for now, which will partially solve this case without needing to figure out runtime semantics for `export`.